File: LeetCode/0238-product-of-array-except-self/solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len = nums.length;
        int[] leftPartialProducts = new int[len];
        int[] rightPartialProducts = new int[len];
        leftPartialProducts[0] = nums[0];
        rightPartialProducts[0] = nums[len - 1];
        for(int i=1;i<len;i++) {
            leftPartialProducts[i] = leftPartialProducts[i - 1] * nums[i];
            rightPartialProducts[i] = rightPartialProducts[i - 1] * nums[len - i - 1];
        }

        int[] result = new int[len];
        for(int k=0;k<len;k++) {
            if(k == 0) {
                result[k] = rightPartialProducts[len - 2];
            }
            else if(k == len - 1) {
                result[k] = leftPartialProducts[len - 2];
            }
            else {
                result[k] = leftPartialProducts[k - 1] * rightPartialProducts[len - k - 2];
            }
        }

        return result;
    }
}

File: LeetCode/0238-product-of-array-except-self/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        std::vector<int> left = { nums[0] };
        std::vector<int> right = { nums[nums.size() - 1] };

        for(int i = 1; i < nums.size();i++) {
            left.push_back(left[i - 1] * nums[i]);
            right.push_back(right[i - 1] * nums[nums.size() - i - 1]);
        }

        std::vector<int> result;
        for(int i=0;i<nums.size();i++) {
            int l = i == 0 ? 1 : left[i - 1];
            int r = i == nums.size() - 1 ? 1 : right[nums.size() - i - 2];
            result.push_back(l * r);
        }

        return result;
    }
};
